src/main/java/org/carlspring/cloud/storage/s3fs/S3FileStore.java

```diff
@@ -107,38 +117,68 @@ public Bucket getBucket()
         return getBucket(name);
     }
 
-    private Bucket getBucket(String bucketName)
+    private Bucket getBucket(final String bucketName)
     {
-        for (Bucket buck : getClient().listBuckets())
+        for (Bucket buck : getClient().listBuckets().buckets())
         {
-            if (buck.getName().equals(bucketName))
+            if (buck.name().equals(bucketName))
             {
                 return buck;
             }
         }
-
         return null;
     }
 
+    private boolean hasBucket(final String bucketName)
+    {
+        // Originally getBucket was being used to determine presence of a bucket
+        //
+        // This is incorrect for two reasons:
+        // 1. The list bucket operation provides buckets for which you are the owner
+        //    It would not, therefore, allow you to work with buckets for which you
+        //    have access but are not the owner.
+        // 2. The way this information is being used later is to determine the
+        //    bucket owner, which by definition, is now "you".
+        // https://docs.aws.amazon.com/AmazonS3/latest/API/RESTServiceGET.html
+        //
+        // However, note that the revised code below now has a different permissions
+        // model as HeadBucket is now required
```

Comment:
Are these notes coming from @elarch 's fork?

Reply:
Yeah, some parts are from this fork that you mention. Could be helpful.

Reply:
The ability to head the bucket is the true test that a bucket "exists" for the Principle as listing buckets only shows those that are in the account of the user whereas "head" will work if the bucket has been shared with the Principle from another account.